gusto/diagnostics.py

```diff
@@ -1687,6 +1687,110 @@ def setup(self, domain, state_fields):
+        vorticity_types = ['relative', 'absolute']
+        if vorticity_type not in vorticity_types:
+            if vorticity_type == 'potential':
+                raise ValueError('Potential vorticity has not yet been implemented')
+            else:
+                raise ValueError(f'vorticity type must be one of {vorticity_types}, not {vorticity_type}')
+
+        space = domain.spaces('HCurl')
+        u = state_fields('u')
+        if self.method != 'solve':
+            if vorticity_type == 'relative':
+                self.expression = curl(u)
+            elif vorticity_type == 'absolute':
+                Omega = as_vector((0, 0, self.parameters.Omega))
+                self.expression = curl(u) + 2*Omega
+
+        super().setup(domain, state_fields, space=space)
+
+        if self.method == 'solve':
+            vort = TrialFunction(space)
+            w = TestFunction(space)
+            n = FacetNormal(domain.mesh)
+            a = inner(vort, w) * dx
+            L = inner(u, curl(w)) * dx - jump(cross(w, u), n) * dS_h
+            if vorticity_type != 'relative':
+                Omega = as_vector((0, 0, self.parameters.Omega))
+                L += inner(2*Omega, w) * dx
+
+            problem = LinearVariationalProblem(a, L, self.field)
+            self.evaluator = LinearVariationalSolver(problem, solver_parameters={'ksp_type': 'cg'})
+
+
+class CompressibleRelativeVorticity(CompressibleVorticity):
+    u""" Diagnostic field for compressible euler relative vorticity  """
+    name = 'CompressibleRelativeVorticity'
+
+    def __init__(self, space=None, method='solve'):
+        u"""
+        Args:
+            space (:class:`FunctionSpace`, optional): the function space to
+                evaluate the diagnostic field in. Defaults to None, in which
+                case a default space will be chosen for this diagnostic.
+            method (str, optional): a string specifying the method of evaluation
+                for this diagnostic. Valid options are 'interpolate', 'project',
+                'assign' and 'solve'. Defaults to 'solve'.
+        """
+        self.solve_implemented = True
+        super().__init__(space=space, method=method, required_fields=('u',))
+
+    def setup(self, domain, state_fields):
+        u"""
+        Sets up the :class:`Function` for the diagnostic field.
+
+        Args:
+            domain (:class:`Domain`): the model's domain object.
+            state_fields (:class:`StateFields`): the model's field container.
+        """
+        super().setup(domain, state_fields, vorticity_type='relative')
+
+
+class CompressibleAbsoluteVorticity(CompressibleVorticity):
+    u""" Diagnostic field for compressible euler absolute vorticity  """
+    name = 'CompressibleAbsoluteVorticity'
+
+    def __init__(self, parameters, space=None, method='solve'):
+        u"""
+        Args:
+            space (:class:`FunctionSpace`, optional): the function space to
```

Comment:
please update docstring to include `parameters` argument

Reply:
I think it's just `CompressibleParameters` - this should also work for the Boussinesq equations so you can remove any mention of Euler... in fact, it also works for incompressible...